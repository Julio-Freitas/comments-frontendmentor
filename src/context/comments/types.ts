export type User = {
  image: {
    png: string;
    webp: string;
  };
  username: string;
};

export type Comment = {
  id: number;
  content: string;
  createdAt: string;
  score: number;
  user: User;
  replies?: Comment[];
};

type ActionScore = 'plus' | 'minus';
export type ActionType = 'comment' | 'reply';

export type ChangeScoreAttr = {
  action: ActionScore;
  type: ActionType;
  id: number;
  commentId?: number | null | undefined;
};

export interface IContext {
  comments: Comment[];
  addNotification?: (value: Comment) => void;
  removeNotification?: (value: Comment) => void;
  changeScoreComment?: ({ type, action, id }: ChangeScoreAttr) => void;
}

export interface IProvider {
  children: React.ReactNode;
}
