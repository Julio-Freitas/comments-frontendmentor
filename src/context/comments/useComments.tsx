import { useContext } from 'react';
import { ContextNotification } from '.';

export function useComments() {
  return useContext(ContextNotification);
}
