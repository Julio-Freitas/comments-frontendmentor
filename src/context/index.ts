export { ContextNotification, ProviderNotification } from './comments';
export { ContextUser, ProviderUser } from './user';
