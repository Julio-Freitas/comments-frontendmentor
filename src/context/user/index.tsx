import { getUser } from './request';
import { IUser } from 'pages/api/user/type';
import { createContext, useEffect, useState } from 'react';

interface ProviderUser {
  children: React.ReactNode;
}
export const ContextUser = createContext<IUser>({} as IUser);

export const ProviderUser = ({ children }: ProviderUser) => {
  const [user, setUser] = useState<IUser | null>();

  useEffect(() => {
    getUser().then(({ data }) => setUser(data));
  }, []);
  return (
    <ContextUser.Provider value={{ ...user } as IUser}>
      {children}
    </ContextUser.Provider>
  );
};
