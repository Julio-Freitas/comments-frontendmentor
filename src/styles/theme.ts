export const Theme = {
  fonts: {
    primary: 'Outfit',
    secondary: 'Plus Jakarta Sans',
  },

  corlors: {
    primary: {
      moderateBlue: 'hsl(238, 40%, 52%)',
      softRed: 'hsl(358, 79%, 66%)',
      lightgrayishBlue: 'hsl(239, 57%, 85%)',
      paleRed: 'hsl(357, 100%, 86%)',
    },
    neutral: {
      darkBlue: ' hsl(212, 24%, 26%)',
      grayishBlue: 'hsl(211, 10%, 45%)',
      lightgray: 'hsl(223, 19%, 93%)',
      veryLightGray: 'hsl(228, 33%, 97%)',
      white: 'hsl(0, 0%, 100%)',
    },
  },
};
