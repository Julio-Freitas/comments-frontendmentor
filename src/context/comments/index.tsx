import { DataResponse } from 'pages/api/comments';
import { createContext, useEffect, useState } from 'react';

import { IContext, IProvider, Comment, ChangeScoreAttr } from './types';

export const ContextNotification = createContext<IContext>({} as IContext);

export const ProviderNotification = ({ children }: IProvider) => {
  const [comments, setComments] = useState<Comment[] | []>([]);

  const getComments = async (): Promise<DataResponse | null> => {
    try {
      const response = await fetch('/api/comments');
      const json = (await response.json()) as DataResponse;
      return json;
    } catch (error) {
      return null;
    }
  };
  useEffect(() => {
    getComments().then((comments) =>
      setComments(comments?.data.comments ?? []),
    );
  }, []);

  const addNotification = (note: Comment) => {
    setComments((prev) => [...prev, note]);
  };

  const removeNotification = (note: Comment) => {
    setComments((prev) => prev.filter((item) => item.id !== note.id));
  };

  const plusScore = (value: number) => (value >= 0 ? ++value : value);
  const minusScore = (value: number) => (value > 0 ? --value : 0);

  const onChangeScoreComment = ({
    action,
    id,
  }: Omit<ChangeScoreAttr, 'type'>) => {
    const currentComment = comments.map((comment) => {
      if (comment.id === id) {
        return {
          ...comment,
          score:
            action === 'plus'
              ? plusScore(comment.score)
              : minusScore(comment.score),
        };
      }
      return comment;
    });

    setComments(currentComment);
  };

  const onChangeScoreReply = ({
    action,
    id,
    commentId,
  }: Omit<ChangeScoreAttr, 'type'>) => {
    const { replies, ...comment } = comments.find(
      (comment) => comment.id === commentId,
    ) as Comment;
    const newScoreReplies = replies?.map((replie) => {
      if (replie.id === id) {
        return {
          ...replie,
          score:
            action === 'plus'
              ? plusScore(replie.score)
              : minusScore(replie.score),
        };
      }
      return replie;
    });
    const newComment = { ...comment, replies: newScoreReplies };
    const currentComments = comments.filter(
      (item) => item.id !== newComment.id,
    );
    setComments([...currentComments, newComment]);
  };

  const changeScoreComment = ({
    type,
    action,
    id,
    commentId,
  }: ChangeScoreAttr) => {
    const actions = {
      comment: () => onChangeScoreComment({ action, id }),
      reply: () => onChangeScoreReply({ action, id, commentId }),
    };

    actions[type]();
  };
  return (
    <ContextNotification.Provider
      value={{
        comments,
        addNotification,
        removeNotification,
        changeScoreComment,
      }}
    >
      {children}
    </ContextNotification.Provider>
  );
};
