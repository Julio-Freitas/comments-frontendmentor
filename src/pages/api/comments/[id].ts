// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import { Comment } from 'context/comments/types';
import type { NextApiRequest, NextApiResponse } from 'next';
import dataString from './data.json';

interface Data {
  data: {
    comment: Comment;
    message: string | null;
  };
}

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>,
) {
  const comments = dataString.comments as Comment[];
  const { id } = req.query;
  const comment =
    (comments.find((comment) => comment.id === Number(id)) as Comment) ?? null;

  const message = !comment ? 'comments not found' : null;
  if (!comment) return res.status(404).json({ data: { comment, message } });

  res.status(200).json({ data: { comment, message } });
}
