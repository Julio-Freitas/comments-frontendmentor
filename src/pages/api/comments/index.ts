// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import { Comment } from 'context/comments/types';
import type { NextApiRequest, NextApiResponse } from 'next';

import data from './data.json';
export interface DataResponse {
  data: {
    comments: Comment[];
  };
}

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<DataResponse>,
) {
  const comments = data.comments as Comment[];
  res.status(200).json({ data: { comments } });
}
