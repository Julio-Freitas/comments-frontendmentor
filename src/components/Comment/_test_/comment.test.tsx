import React from 'react';
import { act, render, screen, waitFor } from '@testing-library/react';
import CommentComponent from '..';
import { Comment } from '../../../context/comments/types';
import { ThemeProvider } from 'styled-components';
import { ProviderNotification } from '../../../context/comments';
import { Theme } from '../../../styles/theme';
import userEvent from '@testing-library/user-event';

const commmentProps: Comment = {
  id: 501,
  content:
    "Impressive! Though it seems the drag feature could be improved. But overall it looks incredible. You've nailed the design and the responsiveness at various breakpoints works really well.",
  createdAt: '1 month ago',
  score: 12,
  user: {
    image: {
      png: '/images/avatars/image-amyrobson.png',
      webp: '/images/avatars/image-amyrobson.webp',
    },
    username: 'amyrobson',
  },
  replies: [],
};
const handleClickMock = jest.fn();

const renderWithProviders = (children: React.ReactNode) => {
  return render(
    <ThemeProvider theme={Theme}>
      <ProviderNotification>{children}</ProviderNotification>
    </ThemeProvider>,
  );
};

describe('<Comment', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('shoul be render correclty', async () => {
    const { container } = await act(async () =>
      renderWithProviders(
        <CommentComponent
          {...commmentProps}
          _handleCLickScore={() => handleClickMock()}
        />,
      ),
    );

    expect(screen.getByText(commmentProps.content)).toBeInTheDocument();
    expect(screen.getByText(commmentProps.createdAt)).toBeInTheDocument();
    expect(screen.getByText(commmentProps.user.username)).toBeInTheDocument();
    expect(screen.getByTestId('reply').firstChild).not.toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be reply if exist', async () => {
    await act(async () => {
      const mockCommentWithReply = {
        ...commmentProps,
        replies: [
          {
            content: 'Content reply',
            createdAt: '10 month ago',
            id: 10,
            score: 1,
            user: {
              image: {
                png: '/images/icon-minus.svg',
                webp: '/images/icon-minus.svg',
              },
              username: 'name reply',
            },
          },
        ] as Comment[],
      };
      renderWithProviders(
        <CommentComponent
          {...mockCommentWithReply}
          _handleCLickScore={() => handleClickMock()}
        />,
      );
    });

    expect(screen.getByText('Content reply')).toBeInTheDocument();
    expect(screen.getByText('10 month ago')).toBeInTheDocument();
    expect(screen.getByText('name reply')).toBeInTheDocument();

    expect(screen.getByTestId('reply').firstChild).toBeInTheDocument();
  });

  it('should be render correclty', async () => {
    await act(async () => {
      renderWithProviders(
        <CommentComponent
          {...commmentProps}
          _handleCLickScore={handleClickMock}
        />,
      );
    });
    const buttonScore = screen.getByTestId('plus');

    await waitFor(() => userEvent.click(buttonScore));

    expect(handleClickMock).toBeCalledTimes(1);
    expect(handleClickMock).toBeCalledWith({
      action: 'plus',
      id: 501,
      type: 'comment',
    });
  });
});
