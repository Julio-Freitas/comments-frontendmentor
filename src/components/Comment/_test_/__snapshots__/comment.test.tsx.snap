// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comment shoul be render correclty 1`] = `
<div
  class="sc-4cc7a831-0 gbxJtS"
>
  <div
    class="sc-4cc7a831-1 hevCLT"
  >
    <div
      class="sc-4cc7a831-2 jxAsaB"
    >
      <button
        class="sc-4cc7a831-3 cExIDm"
        data-testid="plus"
        type="button"
      >
        <img
          alt="Image user"
          data-nimg="1"
          decoding="async"
          height="10"
          loading="lazy"
          src="/images/icon-plus.svg"
          srcset="/images/icon-plus.svg 1x, /images/icon-plus.svg 2x"
          style="color: transparent;"
          width="10"
        />
      </button>
      <p
        class="sc-4cc7a831-4 iutNsS"
        data-testid="score"
      >
        12
      </p>
      <button
        class="sc-4cc7a831-3 cExIDm"
        type="button"
      >
        <img
          alt="Image user"
          data-nimg="1"
          decoding="async"
          height="4"
          loading="lazy"
          src="/images/icon-minus.svg"
          srcset="/images/icon-minus.svg 1x, /images/icon-minus.svg 2x"
          style="color: transparent;"
          width="10"
        />
      </button>
    </div>
    <div
      class="sc-4cc7a831-7 gDUcGA"
    >
      <header
        class="sc-4cc7a831-5 hmBqGq"
      >
        <div
          class="sc-4cc7a831-6 cQyOUm"
        >
          <img
            alt="Image user"
            data-nimg="1"
            decoding="async"
            height="35"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Favatars%2Fimage-amyrobson.webp&w=96&q=75"
            srcset="/_next/image?url=%2Fimages%2Favatars%2Fimage-amyrobson.webp&w=48&q=75 1x, /_next/image?url=%2Fimages%2Favatars%2Fimage-amyrobson.webp&w=96&q=75 2x"
            style="color: transparent;"
            width="35"
          />
          <p>
            amyrobson
          </p>
          <span
            class="createdAt"
          >
            1 month ago
          </span>
        </div>
        <button
          class="sc-4cc7a831-8 jWNuRH"
          type="button"
        >
          Reply
        </button>
      </header>
      <p
        class="sc-4cc7a831-9 cAMEJv"
      >
        Impressive! Though it seems the drag feature could be improved. But overall it looks incredible. You've nailed the design and the responsiveness at various breakpoints works really well.
      </p>
    </div>
  </div>
  <div
    class="sc-4cc7a831-10 dFhxZq"
    data-testid="reply"
  />
</div>
`;
